### ISSUE TITLE
Disclaimer Banner before XRDP login screen

### ISSUE THREAD

**Issue by Criticize** [neutrinolabs/xrdp#547]
Is it possible to create a login/disclaimer banner with a message that the user has to be agree upon before the xrdp credentials prompt appears??

Thank you kindly for any assistance you can lend!

**Comment by Natureshadow** [comment]
It would probably be best to just render `/etc/issue{,.net}` somewhere.

**Comment by proski** [comment]
You can also replace the bitmap with a one containing the disclaimer text.

**Comment by Criticize** [comment]
Currently I have the disclaimer placed in /etc/issue{,.net} and it works perfectly when logging into CentOS via VMware console. However when I use XRDP from a thin-client it skips the disclaimer completely and goes right to the desktop.

Is it possible to trigger a pop-up before the XRDP login screen? OR set XRDP up so that it connects to the normal blue gnome login screen (as if you physically accessing the computer)?

**Comment by Kaydub00** [comment]
Yeah, I can't find a solution to this issue either, so here's what I'm doing. (which now that I just finished writing this, it's basically what @proski said)

I changed the login screen image and title. The image contains the message I need (in my case fedramp related message). The user has to enter their username and password and click okay, which is an implicit agreement to the terms in the image.

To make this work though, the user has to first fail the login or not pass login credentials... so yeah, still not a good solution but it's the only way I could get a message to display for XRDP on CentOS 7.2

**Comment by kfiresmith** [comment]
+1 for a proper standard way to render /etc/issue w/o any BMP logo workarounds

**Comment by taylorjtamu** [comment]
+1 for a way to display /etc/issue without image workarounds.

**Comment by Nexarian** [comment]
I know of someone who would use XRDP if this feature were implemented. I know little about this part of the codebase, but adoption to XRDP is important, so this is now my next highest priority to get implemented... Working on CentOS

**Comment by Nexarian** [comment]
One of the steps is to disable autologin even if the credentials are provided. I suspect we need to add a config for that: https://github.com/neutrinolabs/xrdp/blob/29ef7f896a831bb5e1053221e0837959599d4c34/libxrdp/xrdp_sec.c#L875

**Comment by matt335672** [comment]
You can still do the autologin - you just need to display the message first in a dialog box. I think that's the hard bit.

**Comment by matt335672** [comment]
@Nexarian - one thing I just remembered - there's code for a help dialog in the login screen module. The code is unused, but it might be useful to get you started maybe.

**Comment by Nexarian** [comment]
Here's a prototype screen. I implemented this a little differently. I avoided the problem of displaying another dialog, by simply tacking it on to the main dialog and adding an "always_show_login_window" parameter. Even if credentials are specified, even if autologin is otherwise enabled, this parameter will FORCE the login window to be shown with the acceptable use policy.

Additionally, it's on the system admin to determine the dimensions of the dialog, the file name where the policy is stored, and the actual text of the policy. It doesn't automatically re-compute the size of the dialog to accommodate the policy, it simply starts drawing it as if it's going to fit.

But we can discuss if this is a solution that would satisfy those who want this. Now that I understand this code better, I could go the route of having a "pre-pop-up" dialog as well. That would likely require adding more states to the state machine, not too bad though.
![Example AUP login screen](https://user-images.githubusercontent.com/3752093/218298751-cd12fcda-4817-490a-bc3e-c2499a466a92.png)

https://github.com/neutrinolabs/xrdp/pull/2535

**Comment by Nexarian** [comment]
Thinking on this more, I think the workflow could use a lot of thought. If you've already specified autologin credentials with your RDP client, the use-case is we want to be able to force this message to be displayed before autologin happens. However, we also want to make sure that users aren't annoyed by having to re-enter their credentials. There's a bunch of ways I think this workflow could be adjusted.

1. For autologin credentials that are specified, always display this message, but pre-populate the textboxes. so all you have to do is click ok and the sign in happens.
2. Pre-test that the login WILL succeed but then don't start the session just yet, and hide the textboxes. The more I think about this, the more problematic it likely is.
3. What was discussed earlier: Have a separate "acceptable use policy" dialog that appears before this login screen at all. That is probably the most optimal, as it gets displaying the AUP out of the workflow for auth.

Thoughts appreciated. Like I said, now that I'm more familiar with the code, I think I can get (3) to work.

**Comment by matt335672** [comment]
My vote would be for (3). You should be able to compute the size of the required dialog from the text pretty easily.

**Comment by Nexarian** [comment]
Haha of course you pick the harder one, will work on it!

Might require a lot of refactoring. The state machine for the login window is pretty crazy right now. It "works" but it doesn't have a lot of clear state machine management.

**Comment by umich-steventf** [comment]
> Here's a prototype screen. I implemented this a little differently. I avoided the problem of displaying another dialog, by simply tacking it on to the main dialog and adding an "always_show_login_window" parameter. Even if credentials are specified, even if autologin is otherwise enabled, this parameter will FORCE the login window to be shown with the acceptable use policy.
> 
> Additionally, it's on the system admin to determine the dimensions of the dialog, the file name where the policy is stored, and the actual text of the policy. It doesn't automatically re-compute the size of the dialog to accommodate the policy, it simply starts drawing it as if it's going to fit.
> 
> But we can discuss if this is a solution that would satisfy those who want this. Now that I understand this code better, I could go the route of having a "pre-pop-up" dialog as well. That would likely require adding more states to the state machine, not too bad though. ![Example AUP login screen](https://user-images.githubusercontent.com/3752093/218298751-cd12fcda-4817-490a-bc3e-c2499a466a92.png)
> 
> #2535

I know this may be a long shot, but is there any documentation on how to add a message to the bottom of the login window?
I'm hoping to have it display the output of a command (probably put into a .txt file) that will show users their port numbers to reconnect to sessions.
